### DISCUSSION TITLE
Change colour on a StandardMaterial

### DISCUSSION THREAD

**Discussion by joebain** [bevyengine/bevy#6907] [Q&A] [answered]
I'm trying to change the colour on a StandardMaterial. I had this working previously by using a query to get the material reference:
```mut query: Query<(&mut Enemy, &mut Handle<StandardMaterial>)>,```
and then swapping the material for a different colour one:
```
 for (mut enemy, mut mat_handle) in &mut query {
    *mat_handle = coloured_mat;
}
```
however I had to make my `PbrBundle` a child of the `Enemy` component because I needed to apply a scale and rotation offset. Now I can't get the material handle through the query and I can't work out how to solve this. In other engines I would expect to be able to keep a reference to the material on my `Enemy` and then set the `base_colour` to another one at runtime. I'm not sure if this is possible with Bevy but I'm getting tripped up by the borrow / move / copy restrictions. (I'm new to Rust too, sorry.)

Edit:
To clarify the other approach I tried, I have my `Enemy` struct with a material handle on it:
```
#[derive(Component)]
pub struct Enemy {
    orbit_speed: f32,
    move_speed: f32,
    material: Handle<StandardMaterial>,
}
```
Then I get the material from the asset resource and set the `base_color`. This compiles but the colour doesn't change when I run it.
```
let enemy_mat = materials.get_mut(&enemy.material);
match enemy_mat {
    Some(m) => {
        m.base_color = new_col;
    }
    None => {}
}
```

**Comment by wilk10** [comment, marked as answer]
The first setup doesn't work any more because now the entity with the `PbrBundle` is a child of `Enemy`, whose `Entity` is stored in the `Children` component. So to access the material handle, you now need two queries:
```rust
fn change_material(
    enemies: Query<&Children, With<Enemy>>,
    mut standard_materials: Query<&mut Handle<StandardMaterial>> ,
) { 
    for children_of_enemy in enemies.iter() {
        for child in children_of_enemy.iter() {
            if let Ok(handle) = standard_materials.get_mut(child) {
                // etc
            }
        }
    }
}
```

In the second case it doesn't work because you need to substitute the `Handle`, not mutate the colour. If you change the colour, you need to add the changed material to materials, get the new handle and assign the handle to the entity with `Enemy`, so it's much better to prepare the materials in advance, save the handles somewhere, like you did inside `Enemy`, and access / assign them when needed.

Hope it was helpful!

**Reply by joebain** [reply]
Thanks, that's helpful. The only thing I don't understand now, if I use that query to get `Children` of an `Enemy` how can I also get a reference to the `Enemy` too? I need to set the material based on the state of the `Enemy`.

**Reply by tim-blackbird** [reply]
```rust
fn change_material(
    enemies: Query<(&Children, &Enemy)>,
    mut standard_materials: Query<&mut Handle<StandardMaterial>> ,
) { 
    for (children, enemy) in &enemies {
        for child in children {
            if let Ok(handle) = standard_materials.get_mut(*child) {
                // etc
            }
        }
    }
}
```
